CII-Best-Practices for Nodejs: Silver level (#955)
nodejs/security-wg: tools/ossf_best_practices/silver_criteria.md

> What is the URL for the project (as a whole)?

https://nodejs.org

> What is the URL for the version control repository (it may be the same as the project URL)?

https://github.com/nodejs/node

## Prerequisites

> The project MUST achieve a passing level badge.

**Met**

## Basic project website content

> The information on how to contribute MUST include the requirements for acceptable contributions (e.g., a reference to any required coding standard). (URL required) 

**Met**
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md

## Project oversight

> The project SHOULD have a legal mechanism where all developers of non-trivial amounts of project software assert that they are legally authorized to make these contributions. The most common and easily-implemented approach for doing this is by using a [Developer Certificate of Origin (DCO)](https://developercertificate.org/), where users add "signed-off-by" in their commits and the project links to the DCO website. However, this MAY be implemented as a Contributor License Agreement (CLA), or other legal mechanism. (URL required)

_Possible answers: Met/Unmet_

> > The project MUST clearly define and document its project governance model (the way it makes decisions, including key roles). (URL required)
> 
> _Possible answers: Met/Unmet_

Author: ```suggestion
> The project MUST clearly define and document its project governance model (the way it makes decisions, including key roles). (URL required)

**Met**
https://github.com/nodejs/node/blob/main/GOVERNANCE.md
```

> > The project MUST adopt a code of conduct and post it in a standard location.
> 
> _Possible answers: Met/Unmet_

Author: The project has a clear [CoC](https://github.com/nodejs/node/blob/main/CODE_OF_CONDUCT.md)

```suggestion
> The project MUST adopt a code of conduct and post it in a standard location.

**Met**
```

Reviewer: ```suggestion
> The project MUST adopt a code of conduct and post it in a standard location.

**Met**
https://github.com/nodejs/node/blob/main/CODE_OF_CONDUCT.md
```

> > The project MUST clearly define and publicly document the key roles in the project and their responsibilities, including any tasks those roles must perform. It MUST be clear who has which role(s), though this might not be documented in the same way. (URL required)
> 
> _Possible answers: Met/Unmet_

Author: Maybe there are other documents that are more detailed, but the Governance is enough 

```suggestion
> The project MUST clearly define and publicly document the key roles in the project and their responsibilities, including any tasks those roles must perform. It MUST be clear who has which role(s), though this might not be documented in the same way. (URL required)

**Met**
https://github.com/nodejs/node/blob/main/GOVERNANCE.md
```

> > The project MUST be able to continue with minimal interruption if any one person dies, is incapacitated, or is otherwise unable or unwilling to continue support of the project. In particular, the project MUST be able to create and close issues, accept proposed changes, and release versions of software, within a week of confirmation of the loss of support from any one individual. This MAY be done by ensuring someone else has any necessary keys, passwords, and legal rights to continue the project. Individuals who run a FLOSS project MAY do this by providing keys in a lockbox and a will providing any needed legal rights (e.g., for DNS names).
> 
> _Possible answers: Met/Unmet_

Author: The project has a very solid way to distribute permissions across teams and individuals that are distributed around the globe too.

```suggestion
> The project MUST be able to continue with minimal interruption if any one person dies, is incapacitated, or is otherwise unable or unwilling to continue support of the project. In particular, the project MUST be able to create and close issues, accept proposed changes, and release versions of software, within a week of confirmation of the loss of support from any one individual. This MAY be done by ensuring someone else has any necessary keys, passwords, and legal rights to continue the project. Individuals who run a FLOSS project MAY do this by providing keys in a lockbox and a will providing any needed legal rights (e.g., for DNS names).

**Met**
```

> The project SHOULD have a "bus factor" of 2 or more. (URL required)

> _Possible answers: Met/Unmet_

Reviewer: ```suggestion
**Met**
```

## Documentation

> The project MUST have a documented roadmap that describes what the project intends to do and not do for at least the next year. (URL required)

_Possible answers: Met/Unmet_

The project MUST include documentation of the architecture (aka high-level design) of the software produced by the project. If the project does not produce software, select "not applicable" (N/A). (URL required)

_Possible answers: Met/Unmet/NA_

> > The project MUST document what the user can and cannot expect in terms of security from the software produced by the project (its "security requirements"). (URL required)
> 
> _Possible answers: Met/Unmet_

Author: In the [Security.md](https://github.com/nodejs/node/blob/main/SECURITY.md) we included the thread model that I believe can cover this requirement

```suggestion
> The project MUST document what the user can and cannot expect in terms of security from the software produced by the project (its "security requirements"). (URL required)

**Met**
```

Reviewer 2: https://github.com/nodejs/node/blob/main/SECURITY.md#the-nodejs-threat-model

Author: Solved in 72337c0aea9317487a155de9b5cf4603ab0ffdce

> _Possible answers: Met/Unmet_

Reviewer: ```suggestion
**Met**
this is URL required so I guess https://github.com/nodejs/node/blob/main/SECURITY.md
```

> > The project MUST provide a "quick start" guide for new users to help them quickly do something with the software. (URL required)
> 
> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
> The project MUST provide a "quick start" guide for new users to help them quickly do something with the software. (URL required)

**Met**
https://nodejs.dev/en/learn/
```

The project MUST make an effort to keep the documentation consistent with the current version of the project results (including software produced by the project). Any known documentation defects making it inconsistent MUST be fixed. If the documentation is generally current, but erroneously includes some older information that is no longer true, just treat that as a defect, then track and fix as usual.

_Possible answers: Met/Unmet/NA_

The project repository front page and/or website MUST identify and hyperlink to any achievements, including this best practices badge, within 48 hours of public recognition that the achievement has been attained. (URL required) 

> _Possible answers: Met/Unmet_

Reviewer: i'm not soure about this one? do we have achievements in the front page?

Author: I can create a PR for it once the #954 is completed. I think we can include it easily in the web

## Accessibility and internationalization

> The project (both project sites and project results) SHOULD follow accessibility best practices so that persons with disabilities can still participate in the project and use the project results where it is reasonable to do so.

> _Possible answers: Met/Unmet/NA_

Reviewer: @ovflowd is this requirement met with the new website (accessibility)?

Author: ![Captura de pantalla 2023-05-17 a las 15 04 34](https://github.com/nodejs/security-wg/assets/5110813/ecbeb660-f763-457e-96ca-17388d045711)

Seems like criteria is met @marco-ippolito 🎉

Reviewer 3: @UlisesGascon could you try that test on https://nodejs.dev?

Yes, this is one of the requirements for the new website to meet a11y standards. But be aware that the current template will change to the one on nodejs.dev

Author: Sure thing @ovflowd!

Here is the report:
<img width="1452" alt="Captura de pantalla 2023-05-18 a las 10 53 19" src="https://github.com/nodejs/security-wg/assets/5110813/829cfbf1-4dc1-4fd8-8f7e-65316324dfed">

Seems like only one page can be scanned as the tool didn't find the `robots.txt` or internal links 🤔 

![Captura de pantalla 2023-05-18 a las 10 54 32](https://github.com/nodejs/security-wg/assets/5110813/0a54c6b4-97b5-4cd0-a440-987a4c77f11b)

Maybe is a bug in the tool, because the links are included

<img width="1208" alt="Captura de pantalla 2023-05-18 a las 10 58 05" src="https://github.com/nodejs/security-wg/assets/5110813/ebec9390-a792-41f5-bd66-13958efd0038">

But in any case, we have a very good scoring 🚀

Author: resolved in a31b29567b4a5eca39e4621557640ed848a8417a

> > The software produced by the project SHOULD be internationalized to enable easy localization for the target audience's culture, region, or language. If internationalization (i18n) does not apply (e.g., the software doesn't generate text intended for end-users and doesn't sort human-readable text), select "not applicable" (N/A)
> 
> _Possible answers: Met/Unmet/NA_

Author: I think is N/A for Nodejs itself, but the documentation and the website has support for i18n

```suggestion
> The software produced by the project SHOULD be internationalized to enable easy localization for the target audience's culture, region, or language. If internationalization (i18n) does not apply (e.g., the software doesn't generate text intended for end-users and doesn't sort human-readable text), select "not applicable" (N/A)

**NA**
```

Reviewer: If the website and docs have localization this should be met @ovflowd

Author: Resolved in 245dd9656bf16b3378a1af3816ef7aeb58b2ac69

## Other

> If the project sites (website, repository, and download URLs) store passwords for authentication of external users, the passwords MUST be stored as iterated hashes with a per-user salt by using a key stretching (iterated) algorithm (e.g., Argon2id, Bcrypt, Scrypt, or PBKDF2). If the project sites do not store passwords for this purpose, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**N/A**
```

# Change Control

## Previous versions

> > The project MUST maintain the most often used older versions of the product or provide an upgrade path to newer versions. If the upgrade path is difficult, the project MUST document how to perform the upgrade (e.g., the interfaces that have changed and detailed suggested steps to help upgrade).
> 
> _Possible answers: Met/Unmet/NA_

Author: This is already in place with the [Release Schedule](https://github.com/nodejs/release#release-schedule) and the Release Notes

```suggestion
> The project MUST maintain the most often used older versions of the product or provide an upgrade path to newer versions. If the upgrade path is difficult, the project MUST document how to perform the upgrade (e.g., the interfaces that have changed and detailed suggested steps to help upgrade).

**Met**
```

Reviewer: probably would be good to at the https://github.com/nodejs/release#release-schedule url under met

# Reporting

## Bug-reporting process

> The project MUST use an issue tracker for tracking individual issues. 

**Met**
https://github.com/nodejs/node/issues

## Vulnerability report process

> > The project MUST give credit to the reporter(s) of all vulnerability reports resolved in the last 12 months, except for the reporter(s) who request anonymity. If there have been no vulnerabilities resolved in the last 12 months, select "not applicable" (N/A). (URL required)
> 
> _Possible answers: Met/Unmet/NA_

Author: This is integrated in Hacker One, [Node.js hacktivity](https://hackerone.com/nodejs/hacktivity)
```suggestion
> The project MUST give credit to the reporter(s) of all vulnerability reports resolved in the last 12 months, except for the reporter(s) who request anonymity. If there have been no vulnerabilities resolved in the last 12 months, select "not applicable" (N/A). (URL required)

**Met**
```

Reviewer 2: Met

> > The project MUST have a documented process for responding to vulnerability reports. (URL required)
> 
> _Possible answers: Met/Unmet_

Author: ```suggestion
> The project MUST have a documented process for responding to vulnerability reports. (URL required)

**Met**
https://github.com/nodejs/node/blob/main/SECURITY.md
```

# Quality

## Coding standards

> > The project MUST identify the specific coding style guides for the primary languages it uses, and require that contributions generally comply with it. (URL required)

Reviewer: **Met**
https://github.com/nodejs/node/blob/3cf05be50e02fd9e6933efda6cae7e21b8b531ea/doc/contributing/cpp-style-guide.md

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
### **Met**
https://github.com/nodejs/node/blob/3cf05be50e02fd9e6933efda6cae7e21b8b531ea/doc/contributing/cpp-style-guide.md
```

> The project MUST automatically enforce its selected coding style(s) if there is at least one FLOSS tool that can do so in the selected language(s).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```

_Possible answers: Met/Unmet/NA_

## Installation system

> The project MUST provide a way to easily install and uninstall the software produced by the project using a commonly-used convention.

_Possible answers: Met/Unmet/NA_

> The installation system for end-users MUST honor standard conventions for selecting the location where built artifacts are written to at installation time. For example, if it installs files on a POSIX system it MUST honor the DESTDIR environment variable. If there is no installation system or no standard convention, select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

> The project MUST provide a way for potential developers to quickly install all the project results and support environment necessary to make changes, including the tests and test environment. This MUST be performed with a commonly-used convention.

_Possible answers: Met/Unmet/NA_

## Externally-maintained components

> The project MUST list external dependencies in a computer-processable way. (URL required)

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/maintaining/maintaining-dependencies.md
```

> Projects MUST monitor or periodically check their external dependencies (including convenience copies) to detect known vulnerabilities, and fix exploitable vulnerabilities or verify them as unexploitable.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```

> The project MUST either:
>   1. make it easy to identify and update reused externally-maintained components; or
>   2. use the standard components provided by the system or programming language.
> 
> Then, if a vulnerability is found in a reused component, it will be easy to update that component.

_Possible answers: Met/Unmet/NA_

> The project SHOULD avoid using deprecated or obsolete functions and APIs where FLOSS alternatives are available in the set of technology it uses (its "technology stack") and to a supermajority of the users the project supports (so that users have ready access to the alternative).

_Possible answers: Met/Unmet/NA_

## Automated test suite

> > An automated test suite MUST be applied on each check-in to a shared repository for at least one branch. This test suite MUST produce a report on test success or failure.
> 
> _Possible answers: Met/Unmet_

Author: ```suggestion
> An automated test suite MUST be applied on each check-in to a shared repository for at least one branch. This test suite MUST produce a report on test success or failure.

**Met**
```

Reviewer 2: Met

Author: Resolved in 0c09b4d8c031febe0d156ba910bc0a8d470d678e

> The project MUST add regression tests to an automated test suite for at least 50% of the bugs fixed within the last six months.

_Possible answers: Met/Unmet/NA_

> The project MUST have FLOSS automated test suite(s) that provide at least 80% statement coverage if there is at least one FLOSS tool that can measure this criterion in the selected language.

_Possible answers: Met/Unmet/NA_

## New functionality testing

> > The project MUST have a formal written policy that as major new functionality is added, tests for the new functionality MUST be added to an automated test suite.
> 
> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
> The project MUST have a formal written policy that as major new functionality is added, tests for the new functionality MUST be added to an automated test suite.

**Met**
```

Reviewer 2: Met

Author: Resolved in 4c286d63e1bfb1a1d3688f2514b37042a3d1d082

> > The project MUST include, in its documented instructions for change proposals, the policy that tests are to be added for major new functionality.
> 
> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
> The project MUST include, in its documented instructions for change proposals, the policy that tests are to be added for major new functionality.

**Met**
```

Author: Resolved in 7b1b8ef1fe3c95d885b9713bd3b7fafa3fdd6f7e

## Warning flags

> Projects MUST be maximally strict with warnings in the software produced by the project, where practical.

_Possible answers: Met/Unmet/NA_

# Security

## Secure development knowledge

> > The project MUST implement secure design principles (from "know_secure_design"), where applicable. If the project is not producing software, select "not applicable" (N/A).
> 
> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
> The project MUST implement secure design principles (from "know_secure_design"), where applicable. If the project is not producing software, select "not applicable" (N/A).

** Met**
```

_Note that some software does not need to use cryptographic mechanisms. If your project produces software that (1) includes, activates, or enables encryption functionality, and (2) might be released from the United States (US) to outside the US or to a non-US-citizen, you may be legally required to take a few extra steps. Typically this just involves sending an email. For more information, see the encryption section of [Understanding Open Source Technology & US Export Controls](https://www.linuxfoundation.org/resources/publications/understanding-us-export-controls-with-os-projects/)._

> The default security mechanisms within the software produced by the project MUST NOT depend on cryptographic algorithms or modes with known serious weaknesses (e.g., the SHA-1 cryptographic hash algorithm or the CBC mode in SSH).

_Possible answers: Met/Unmet/NA_

> The project SHOULD support multiple cryptographic algorithms, so users can quickly switch if one is broken. Common symmetric key algorithms include AES, Twofish, and Serpent. Common cryptographic hash algorithm alternatives include SHA-2 (including SHA-224, SHA-256, SHA-384 AND SHA-512) and SHA-3.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```
I'm not really sure if it is applicable tho

Author: True, this one is tricky. I am quite sure is `met` as we provide a clear API with several crypto algorithms that the user can interchange easily if one of them is compromised.

> The project MUST support storing authentication credentials (such as passwords and dynamic tokens) and private cryptographic keys in files that are separate from other information (such as configuration files, databases, and logs), and permit users to update and replace them without code recompilation. If the project never processes authentication credentials and private cryptographic keys, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**N/A**
```
we dont store passwords

_Possible answers: Met/Unmet/NA_

> The software produced by the project SHOULD, if it supports or uses TLS, support at least TLS version 1.2. Note that the predecessor of TLS was called SSL. If the software does not use TLS, select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

> The software produced by the project MUST, if it supports TLS, perform TLS certificate verification by default when using TLS, including on subresources. If the software does not use TLS, select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

> The software produced by the project MUST, if it supports TLS, perform certificate verification before sending HTTP headers with private information (such as secure cookies). If the software does not use TLS, select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

## Secure release

> > The project MUST cryptographically sign releases of the project results intended for widespread use, and there MUST be a documented process explaining to users how they can obtain the public signing keys and verify the signature(s). The private key for these signature(s) MUST NOT be on site(s) used to directly distribute the software to the public. If releases are not intended for widespread use, select "not applicable" (N/A).

Reviewer 2: Met

> > The project MUST cryptographically sign releases of the project results intended for widespread use, and there MUST be a documented process explaining to users how they can obtain the public signing keys and verify the signature(s). The private key for these signature(s) MUST NOT be on site(s) used to directly distribute the software to the public. If releases are not intended for widespread use, select "not applicable" (N/A).

Author: Resolved in 31a69d5b5c4e7b3061bfd2ba292c006f793cf69c

> > The project MUST cryptographically sign releases of the project results intended for widespread use, and there MUST be a documented process explaining to users how they can obtain the public signing keys and verify the signature(s). The private key for these signature(s) MUST NOT be on site(s) used to directly distribute the software to the public. If releases are not intended for widespread use, select "not applicable" (N/A).

Reviewer 4: It may be worth clarifying some nuance here. The project only signs the [`SHASUMS.txt`](https://nodejs.org/dist/latest-v20.x/SHASUMS256.txt) file rather than signing each of the release assets.

_Possible answers: Met/Unmet/NA_

> The project MUST have a documented process for responding to vulnerability reports. (URL required)

_Possible answers: Met/Unmet_

# Quality

## Coding standards

> The project MUST identify the specific coding style guides for the primary languages it uses, and require that contributions generally comply with it. (URL required)

**Met**
### **Met**
https://github.com/nodejs/node/blob/main/doc/contributing/cpp-style-guide.md

> The project MUST automatically enforce its selected coding style(s) if there is at least one FLOSS tool that can do so in the selected language(s).

**Met**

## Working build system

> Build systems for native binaries MUST honor the relevant compiler and linker (environment) variables passed in to them (e.g., CC, CFLAGS, CXX, CXXFLAGS, and LDFLAGS) and pass them to compiler and linker invocations. A build system MAY extend them with additional flags; it MUST NOT simply replace provided values with its own. If no native binaries are being generated, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```

> The build and installation system SHOULD preserve debugging information if they are requested in the relevant flags (e.g., "install -s" is not used). If there is no build or installation system (e.g., typical JavaScript libraries), select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```
we have debug flags for build

> The build system for the software produced by the project MUST NOT recursively build subdirectories if there are cross-dependencies in the subdirectories. If there is no build or installation system (e.g., typical JavaScript libraries), select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

> The project MUST be able to repeat the process of generating information from source files and get exactly the same bit-for-bit result. If no building occurs (e.g., scripting languages where the source code is used directly instead of being compiled), select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

## Installation system

> The project MUST provide a way to easily install and uninstall the software produced by the project using a commonly-used convention.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```
we have installers https://nodejs.org/en

> The installation system for end-users MUST honor standard conventions for selecting the location where built artifacts are written to at installation time. For example, if it installs files on a POSIX system it MUST honor the DESTDIR environment variable. If there is no installation system or no standard convention, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```
it is possible to customize installation folder

> The project MUST provide a way for potential developers to quickly install all the project results and support environment necessary to make changes, including the tests and test environment. This MUST be performed with a commonly-used convention.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#setting-up-your-local-environment
```

## Externally-maintained components

> The project MUST list external dependencies in a computer-processable way. (URL required)

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/maintaining/maintaining-dependencies.md

> Projects MUST monitor or periodically check their external dependencies (including convenience copies) to detect known vulnerabilities, and fix exploitable vulnerabilities or verify them as unexploitable.

**Met**

> The project MUST either:
>   1. make it easy to identify and update reused externally-maintained components; or
>   2. use the standard components provided by the system or programming language.
> 
> Then, if a vulnerability is found in a reused component, it will be easy to update that component.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
```
We have update scripts for dependencies

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/releases.md#11-tag-and-sign-the-release-commit

## Other security issues

> The project results MUST check all inputs from potentially untrusted sources to ensure they are valid (an *allowlist*), and reject invalid inputs, if there are any restrictions on the data at all.

_Possible answers: Met/Unmet/NA_

> Hardening mechanisms SHOULD be used in the software produced by the project so that software defects are less likely to result in security vulnerabilities.

_Possible answers: Met/Unmet/NA_

> The project MUST provide an assurance case that justifies why its security requirements are met. The assurance case MUST include: a description of the threat model, clear identification of trust boundaries, an argument that secure design principles have been applied, and an argument that common implementation security weaknesses have been countered. (URL required)

> _Possible answers: Met/Unmet_

Reviewer: ```suggestion
**Met**
https://github.com/nodejs/node/blob/main/SECURITY.md
```

_Possible answers: Met/Unmet/NA_

> The project MUST document what the user can and cannot expect in terms of security from the software produced by the project (its "security requirements"). (URL required)

**Met**
https://github.com/nodejs/node/blob/main/SECURITY.md

> The project MUST provide a "quick start" guide for new users to help them quickly do something with the software. (URL required)

**Met**
https://nodejs.dev/en/learn/

The project MUST make an effort to keep the documentation consistent with the current version of the project results (including software produced by the project). Any known documentation defects making it inconsistent MUST be fixed. If the documentation is generally current, but erroneously includes some older information that is no longer true, just treat that as a defect, then track and fix as usual.

_Possible answers: Met/Unmet/NA_

The project repository front page and/or website MUST identify and hyperlink to any achievements, including this best practices badge, within 48 hours of public recognition that the achievement has been attained. (URL required) 

> _Possible answers: Met/Unmet_

Reviewer: ```suggestion
**Met**
```
we have to be fast on this one 😂 only 48 after we get first badge

_Possible answers: Met/Unmet/NA_

> The software produced by the project SHOULD be internationalized to enable easy localization for the target audience's culture, region, or language. If internationalization (i18n) does not apply (e.g., the software doesn't generate text intended for end-users and doesn't sort human-readable text), select "not applicable" (N/A)

_Possible answers: Met/Unmet/NA_

## Other

> If the project sites (website, repository, and download URLs) store passwords for authentication of external users, the passwords MUST be stored as iterated hashes with a per-user salt by using a key stretching (iterated) algorithm (e.g., Argon2id, Bcrypt, Scrypt, or PBKDF2). If the project sites do not store passwords for this purpose, select "not applicable" (N/A).

**N/A**

# Change Control

## Previous versions

> The project MUST maintain the most often used older versions of the product or provide an upgrade path to newer versions. If the upgrade path is difficult, the project MUST document how to perform the upgrade (e.g., the interfaces that have changed and detailed suggested steps to help upgrade).

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
https://github.com/nodejs/release#release-schedule
```

## Secure release

> The project MUST cryptographically sign releases of the project results intended for widespread use, and there MUST be a documented process explaining to users how they can obtain the public signing keys and verify the signature(s). The private key for these signature(s) MUST NOT be on site(s) used to directly distribute the software to the public. If releases are not intended for widespread use, select "not applicable" (N/A).

_Possible answers: Met/Unmet/NA_

> It is SUGGESTED that in the version control system, each important version tag (a tag that is part of a major release, minor release, or fixes publicly noted vulnerabilities) be cryptographically signed and verifiable as described in [signed_releases](https://bestpractices.coreinfrastructure.org/en/projects/29?criteria_level=1#signed_releases).

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/releases.md#11-tag-and-sign-the-release-commit

## Other security issues

> The project results MUST check all inputs from potentially untrusted sources to ensure they are valid (an *allowlist*), and reject invalid inputs, if there are any restrictions on the data at all.

> _Possible answers: Met/Unmet/NA_

Reviewer: This question is quite generic but it's met since we validate inputs

```suggestion
**Met**
```

> Hardening mechanisms SHOULD be used in the software produced by the project so that software defects are less likely to result in security vulnerabilities.

_Possible answers: Met/Unmet/NA_

> The project MUST provide an assurance case that justifies why its security requirements are met. The assurance case MUST include: a description of the threat model, clear identification of trust boundaries, an argument that secure design principles have been applied, and an argument that common implementation security weaknesses have been countered. (URL required)

_Possible answers: Met/Unmet_

# Analysis

## Static code analysis

> The project MUST use at least one static analysis tool with rules or approaches to look for common vulnerabilities in the analyzed language or environment, if there is at least one FLOSS tool that can implement this criterion in the selected language.

> _Possible answers: Met/Unmet/NA_

Reviewer: ```suggestion
**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/static-analysis.md
```
I guess this is still a thing?

**Met**
https://github.com/nodejs/node/blob/main/GOVERNANCE.md

> The project MUST be able to continue with minimal interruption if any one person dies, is incapacitated, or is otherwise unable or unwilling to continue support of the project. In particular, the project MUST be able to create and close issues, accept proposed changes, and release versions of software, within a week of confirmation of the loss of support from any one individual. This MAY be done by ensuring someone else has any necessary keys, passwords, and legal rights to continue the project. Individuals who run a FLOSS project MAY do this by providing keys in a lockbox and a will providing any needed legal rights (e.g., for DNS names).

**Met**

> The project SHOULD have a "bus factor" of 2 or more. (URL required)

**Met**
https://github.com/nodejs/node/blob/main/README.md#current-project-team-members

## Documentation

> The project MUST have a documented roadmap that describes what the project intends to do and not do for at least the next year. (URL required)

> _Possible answers: Met/Unmet_

Reviewer 2: I think we can use https://github.com/nodejs/node/blob/HEAD/doc/contributing/strategic-initiatives.md

> _Possible answers: Met/Unmet_

Author: Solved in a83a6e9

The project MUST include documentation of the architecture (aka high-level design) of the software produced by the project. If the project does not produce software, select "not applicable" (N/A). (URL required)

> _Possible answers: Met/Unmet/NA_

Reviewer 2: Maybe https://github.com/nodejs/node/tree/main/doc/contributing?

> _Possible answers: Met/Unmet/NA_

Reviewer 5: I agree with @RafaelGSS that pointing to https://github.com/nodejs/node/tree/main/doc/contributing makes sense.

> _Possible answers: Met/Unmet/NA_

Author: Solved in 582d364

> The project MUST document what the user can and cannot expect in terms of security from the software produced by the project (its "security requirements"). (URL required)

**Met**
https://github.com/nodejs/node/blob/main/SECURITY.md and https://github.com/nodejs/node/blob/main/SECURITY.md#the-nodejs-threat-model

> The project MUST provide a "quick start" guide for new users to help them quickly do something with the software. (URL required)

**Met**
https://nodejs.dev/en/learn/

The project MUST make an effort to keep the documentation consistent with the current version of the project results (including software produced by the project). Any known documentation defects making it inconsistent MUST be fixed. If the documentation is generally current, but erroneously includes some older information that is no longer true, just treat that as a defect, then track and fix as usual.

> _Possible answers: Met/Unmet/NA_

Reviewer 2: Met

The project repository front page and/or website MUST identify and hyperlink to any achievements, including this best practices badge, within 48 hours of public recognition that the achievement has been attained. (URL required) 

> **Met**

Reviewer 6: This appears to be missing the required URL.

Author: Yes, good catch @tniessen. We have an issue open for it: https://github.com/nodejs/nodejs.org/issues/5432#issuecomment-1687553279. I hope that we can integrate it in the website 🤞

> The project MUST automatically enforce its selected coding style(s) if there is at least one FLOSS tool that can do so in the selected language(s).

**Met**

## Working build system

> Build systems for native binaries MUST honor the relevant compiler and linker (environment) variables passed in to them (e.g., CC, CFLAGS, CXX, CXXFLAGS, and LDFLAGS) and pass them to compiler and linker invocations. A build system MAY extend them with additional flags; it MUST NOT simply replace provided values with its own. If no native binaries are being generated, select "not applicable" (N/A).

**Met**

> The build and installation system SHOULD preserve debugging information if they are requested in the relevant flags (e.g., "install -s" is not used). If there is no build or installation system (e.g., typical JavaScript libraries), select "not applicable" (N/A).

**Met**

> The build system for the software produced by the project MUST NOT recursively build subdirectories if there are cross-dependencies in the subdirectories. If there is no build or installation system (e.g., typical JavaScript libraries), select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer 2: I think we don't have are cross-dependencies build. We do recursively build, but there are no different versions of the same dependency in the project. 

Am I correct? @nodejs/build

> _Possible answers: Met/Unmet/NA_

Reviewer 7: > there are no different versions of the same dependency in the project.

There are -- V8 includes a copy of zlib which is not the same as the one we have in `deps/zlib`. (There is a PR open (https://github.com/nodejs/node/pull/47493) to see if we can get rid of one of them.)

I don't think that counts as a cross-dependency though.

> _Possible answers: Met/Unmet/NA_

Reviewer: I think this is met, we do not recursively build

> _Possible answers: Met/Unmet/NA_

Author: Closed by 98fcafb6dcde9b13f31583ec5aa93b7e604056ff

> The project MUST be able to repeat the process of generating information from source files and get exactly the same bit-for-bit result. If no building occurs (e.g., scripting languages where the source code is used directly instead of being compiled), select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Reviewer 7: `exactly the same bit-for-bit result` sounds like reproducible builds which we don't currently have.

> _Possible answers: Met/Unmet/NA_

Reviewer 2: @richardlau Wouldn't the promoted binaries https://ci-release.nodejs.org/job/iojs+release/ be the reproducible builds?

> _Possible answers: Met/Unmet/NA_

Reviewer 7: @RafaelGSS no, because if you rebuilt them they would not be bit-for-bit identical.

> _Possible answers: Met/Unmet/NA_

Reviewer 5: @sxa has been looking at reproduciable builds so supporting his work may help the project achieve this requirement.

> _Possible answers: Met/Unmet/NA_

Author: this is unmet for now.

> _Possible answers: Met/Unmet/NA_

Author: This has been completed in f6dbb3f5d6c65c5803a21ce428bae358b4c07598, waiting for the issue tracking in to resolve the conversation.

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#setting-up-your-local-environment

## Externally-maintained components

> The project MUST list external dependencies in a computer-processable way. (URL required)

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/maintaining/maintaining-dependencies.md

> Projects MUST monitor or periodically check their external dependencies (including convenience copies) to detect known vulnerabilities, and fix exploitable vulnerabilities or verify them as unexploitable.

**Met**

> The project MUST either:
>   1. make it easy to identify and update reused externally-maintained components; or
>   2. use the standard components provided by the system or programming language.
> 
> Then, if a vulnerability is found in a reused component, it will be easy to update that component.

**Met**

> The project SHOULD avoid using deprecated or obsolete functions and APIs where FLOSS alternatives are available in the set of technology it uses (its "technology stack") and to a supermajority of the users the project supports (so that users have ready access to the alternative).

> _Possible answers: Met/Unmet/NA_

Reviewer 2: Met

## Automated test suite

> An automated test suite MUST be applied on each check-in to a shared repository for at least one branch. This test suite MUST produce a report on test success or failure.

**Met**

> The project MUST add regression tests to an automated test suite for at least 50% of the bugs fixed within the last six months.

> _Possible answers: Met/Unmet/NA_

Reviewer 2: Met, yeah.

> _Possible answers: Met/Unmet/NA_

Reviewer 5: I agree that we add tests for bugs fixed.

> _Possible answers: Met/Unmet/NA_

Author: Solved in 6d842e1

> The project MUST have FLOSS automated test suite(s) that provide at least 80% statement coverage if there is at least one FLOSS tool that can measure this criterion in the selected language.

> _Possible answers: Met/Unmet/NA_

Reviewer 7: FYI https://app.codecov.io/gh/nodejs/node 🙂

> _Possible answers: Met/Unmet/NA_

Author: Completed in 8ba7500

## New functionality testing

> The project MUST have a formal written policy that as major new functionality is added, tests for the new functionality MUST be added to an automated test suite.

**Met**

> The project MUST include, in its documented instructions for change proposals, the policy that tests are to be added for major new functionality.

**Met**

## Warning flags

> Projects MUST be maximally strict with warnings in the software produced by the project, where practical.

> _Possible answers: Met/Unmet/NA_

Reviewer 5: We'd have to dig it out of the build scripts or jobs but I believe we have enable errors on warnings for most cases.

> _Possible answers: Met/Unmet/NA_

Author: Confirmed, this is met

> _Possible answers: Met/Unmet/NA_

Author: Resolved in c5b47414a0e87f0bd10c004fb1cc8d694296564d

# Security

## Secure development knowledge

> The project MUST implement secure design principles (from "know_secure_design"), where applicable. If the project is not producing software, select "not applicable" (N/A).

** Met**

## Use basic good cryptographic practices

_Note that some software does not need to use cryptographic mechanisms. If your project produces software that (1) includes, activates, or enables encryption functionality, and (2) might be released from the United States (US) to outside the US or to a non-US-citizen, you may be legally required to take a few extra steps. Typically this just involves sending an email. For more information, see the encryption section of [Understanding Open Source Technology & US Export Controls](https://www.linuxfoundation.org/resources/publications/understanding-us-export-controls-with-os-projects/)._

> The default security mechanisms within the software produced by the project MUST NOT depend on cryptographic algorithms or modes with known serious weaknesses (e.g., the SHA-1 cryptographic hash algorithm or the CBC mode in SSH).

> _Possible answers: Met/Unmet/NA_

Author: I think is met, but a lot depends on how the final users will build the applications (ex. md5 to store a pass, etc..). Maybe is `N/A`?

Reviewer 2: @tniessen any thoughts?

Reviewer: I think we provide those algorithms to users, but dont rely on them

Author: Confirmed, this is met. We don't use weak algorithms internally.

Author: resolved in 80520539e11cc910ef838962808e7023f514f7f6

**Met**

> The project MUST support storing authentication credentials (such as passwords and dynamic tokens) and private cryptographic keys in files that are separate from other information (such as configuration files, databases, and logs), and permit users to update and replace them without code recompilation. If the project never processes authentication credentials and private cryptographic keys, select "not applicable" (N/A).

**N/A**

> The software produced by the project SHOULD support secure protocols for all of its network communications, such as SSHv2 or later, TLS1.2 or later (HTTPS), IPsec, SFTP, and SNMPv3. Insecure protocols such as FTP, HTTP, telnet, SSLv3 or earlier, and SSHv1 SHOULD be disabled by default, and only enabled if the user specifically configures it. If the software produced by the project does not support network communications, select "not applicable" (N/A). 

> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
**Met**
```

Reviewer 5: I think this could be N/A as by default there is no communicate endpoint enabled.

Author: Solved in 93bbe44

> The software produced by the project SHOULD, if it supports or uses TLS, support at least TLS version 1.2. Note that the predecessor of TLS was called SSL. If the software does not use TLS, select "not applicable" (N/A).

**Met**

> The software produced by the project MUST, if it supports TLS, perform TLS certificate verification by default when using TLS, including on subresources. If the software does not use TLS, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
**Met**
```
Can you confirm @RafaelGSS ?

Reviewer 2: Yeah, we do have TLS validation, not sure if that's performed by default, but I think so

Reviewer 5: We'd need to confirm

Reviewer 6: TLS certificate verification is done by default by all relevant APIs.

> The software produced by the project MUST, if it supports TLS, perform certificate verification before sending HTTP headers with private information (such as secure cookies). If the software does not use TLS, select "not applicable" (N/A).

> _Possible answers: Met/Unmet/NA_

Author: ```suggestion
**Met**
```
Can you confirm @RafaelGSS ?

Reviewer 2: cc: @shogunpanda

Reviewer 8: I might have to double check this but I think the answer is yes.
I'm 99% we do this in the `tls` module, which is internally used by `https`.

@mcollina @jasnell do you confirm?

**Met** https://github.com/nodejs/node/blob/main/doc/contributing/releases.md#3-a-publicly-listed-gpg-key

> It is SUGGESTED that in the version control system, each important version tag (a tag that is part of a major release, minor release, or fixes publicly noted vulnerabilities) be cryptographically signed and verifiable as described in [signed_releases](https://bestpractices.coreinfrastructure.org/en/projects/29?criteria_level=1#signed_releases).

**Met**
https://github.com/nodejs/node/blob/main/doc/contributing/releases.md#11-tag-and-sign-the-release-commit

## Other security issues

> The project results MUST check all inputs from potentially untrusted sources to ensure they are valid (an *allowlist*), and reject invalid inputs, if there are any restrictions on the data at all.

**Met**

> Hardening mechanisms SHOULD be used in the software produced by the project so that software defects are less likely to result in security vulnerabilities.

> _Possible answers: Met/Unmet/NA_

Author: Not sure if this apply to Node 🤔

Reviewer 2: Shouldn't it be the isolated build machines?

Reviewer 5: Is there a definition of `Hardening mechanisms`, I don't think the use of isolated build machines is `in the software`

Reviewer 6: I assume this refers to hardening mechanisms such as `-fsanitize=bounds`, `-fstack-protector`, etc. Some are enabled by default but most come with significant performance overheads, thus disabled.

Author: So, based on [this](https://github.com/coreinfrastructure/best-practices-badge/blob/f4d207721526c1f60c2003f747a46ae897ad9977/doc/other.md?plain=1#L1070):

> Hardening mechanisms may include HTTP headers like Content Security Policy (CSP), compiler flags to mitigate attacks (such as -fstack-protector), or compiler flags to eliminate undefined behavior. For our purposes least privilege is not considered a hardening mechanism (least privilege is important, but separate).

I think this is a `N/A` in our case, because the final user has the power to decide. As well we clarified this in [our thread model](https://github.com/nodejs/node/blob/main/SECURITY.md#the-nodejs-threat-model)

Reviewer 6: > I think this is a N/A in our case, because the final user has the power to decide.

@UlisesGascon Not all of the mentioned examples are up to the user, e.g., "compiler flags to mitigate attacks (such as -fstack-protector), or compiler flags to eliminate undefined behavior" (which is what I referred to in https://github.com/nodejs/security-wg/pull/955#discussion_r1255662698).

Reviewer 2: Maybe N/A and mention those cases?

Author: > Maybe N/A and mention those cases?

Agree, do we have a list of the relevant flags so we can include a URL or similar?

Reviewer 6: I don't know any specific list. The exact compiler flags likely vary depending on the target operating system, the compiler, the build type (debug/release/asan/...), etc.

Reviewer 2: What about:

> N/A - However, specific compiler flags such as `-fstack-protector` might not be passed to the distributed binary

Also, might be worth it to check other runtimes answer (if they have this initiative)

Author: Quick update: In the last team meeting #1059 we agreed to use "N/A". We can review it again in the next iteration if needed :)
